gen/template.go

```diff
@@ -1,93 +1,94 @@
-func applyTemplate(f *descriptor.File, opts Options) (string, error) {
-	w := bytes.NewBuffer(nil)
+func ApplyTemplate(w io.Writer, f *protogen.File, opts Options) error {
 
 	if err := headerTemplate.Execute(w, tplHeader{
 		File: f,
 	}); err != nil {
-		return "", err
+		return err
 	}
 
-	for _, msg := range f.Messages {
-		glog.V(2).Infof("Processing %s", msg.GetName())
-		if msg.Options != nil && msg.Options.GetMapEntry() {
-			glog.V(2).Infof("Skipping %s, mapentry message", msg.GetName())
+	return applyMessages(w, f.Messages, opts)
+}
+
+func applyMessages(w io.Writer, msgs []*protogen.Message, opts Options) error {
+	for _, m := range msgs {
+
+		if m.Desc.IsMapEntry() {
+			glog.V(2).Infof("Skipping %s, mapentry message", m.GoIdent.GoName)
 			continue
 		}
-		msgName := camelCase(*msg.Name)
-		msg.Name = &msgName
+
+		glog.V(2).Infof("Processing %s", m.GoIdent.GoName)
 		if err := messageTemplate.Execute(w, tplMessage{
-			Message: msg,
+			Message: m,
 			Options: opts,
 		}); err != nil {
-			return "", err
+			return err
+		}
+
+		if err := applyMessages(w, m.Messages, opts); err != nil {
+			return err
 		}
+
 	}
 
-	return w.String(), nil
+	return nil
 }
 
 type tplHeader struct {
-	*descriptor.File
+	*protogen.File
 }
 
 type tplMessage struct {
-	*descriptor.Message
+	*protogen.Message
 	Options
 }
 
-// TypeName returns the name of the type for this message. This logic
-// is based on the logic of Descriptor.TypeName in golang/protobuf.
-func (t tplMessage) TypeName() string {
-	if len(t.Outers) > 0 {
-		return strings.Join(t.Outers, "_") + "_" + *t.Name
-	}
-
-	return *t.Name
-}
-
 var (
 	headerTemplate = template.Must(template.New("header").Parse(`
 // Code generated by protoc-gen-go-json. DO NOT EDIT.
-// source: {{.GetName}}
+// source: {{.Proto.Name}}
 
-package {{.GoPkg.Name}}
+package {{.GoPackageName}}
 
 import (
-	"bytes"
-
-	"github.com/golang/protobuf/jsonpb"
+	"google.golang.org/protobuf/encoding/protojson"
 )
 `))
 
 	messageTemplate = template.Must(template.New("message").Parse(`
 // MarshalJSON implements json.Marshaler
-func (msg *{{.TypeName}}) MarshalJSON() ([]byte,error) {
-	var buf bytes.Buffer
-	err := (&jsonpb.Marshaler{
-	  EnumsAsInts: {{.EnumsAsInts}},
-	  EmitDefaults: {{.EmitDefaults}},
-	  OrigName: {{.OrigName}},
-	}).Marshal(&buf, msg)
-	return buf.Bytes(), err
+func (msg *{{.GoIdent.GoName}}) MarshalJSON() ([]byte,error) {
+	return protojson.MarshalOptions {
+		UseEnumNumbers: {{.EnumsAsInts}},
+		EmitUnpopulated: {{.EmitDefaults}},
+		UseProtoNames: {{.OrigName}},
+	}.Marshal(msg)
 }
 
 // UnmarshalJSON implements json.Unmarshaler
-func (msg *{{.TypeName}}) UnmarshalJSON(b []byte) error {
-	return (&jsonpb.Unmarshaler{
-	  AllowUnknownFields: {{.AllowUnknownFields}},
-	}).Unmarshal(bytes.NewReader(b), msg)
+func (msg *{{.GoIdent.GoName}}) UnmarshalJSON(b []byte) error {
+	return protojson.UnmarshalOptions {
+		DiscardUnknown: {{.AllowUnknownFields}},
```

Comment:
I think this should be `{{not .AllowUnknownFields}},`

Reply:
According to the protojson docs:

> 	// If DiscardUnknown is set, unknown fields are ignored.
	DiscardUnknown bool

The field controlling the functionality used to be called `AllowUnknownFields` in jsonpb.

> 	// AllowUnknownFields specifies whether to allow messages to contain
	// unknown JSON fields, as opposed to failing to unmarshal.
	AllowUnknownFields bool

But the [behavior is the same](https://github.com/golang/protobuf/blob/4846b58453b3708320bdb524f25cc5a1d9cda4d4/jsonpb/decode.go#L118-L121).

Reply:
Confusing, but makes sense. Thanks for clarifying.

Reply:
Yeah, I'm not sure why they changed the field name. I prefer the old one myself. 😄